Create New Project contrib and store project configuration

sharon-wang opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://connect.posit.it/positron-wiki/pull-requests.html
* Ensure that the code is up-to-date with the `main` branch.
-->

### Intent

- Most of this work is scaffolding for #2818, #2819, #2820
- Address https://github.com/posit-dev/positron/issues/2939
    - related to https://github.com/posit-dev/positron/issues/2878

#### Summary
- scaffolding to set up New Project contrib and service
- storing the new project config in the storage service
- some cleanup and refactoring to ensure the project config is stored correctly
- adding config setting to support testing the project wizard work in release builds

#### Config Setting Preview
<img width="710" alt="image" src="https://github.com/posit-dev/positron/assets/25834218/0a57231e-72b9-4e0a-b505-e466c215420e">

### Approach

- store the project configuration in the storage service
    - create interface for this data so it can be JSON stringified on write and JSON parsed on read
    - use `StorageScope.APPLICATION` so any workspace can check to see if it is the new project workspace and `StorageTarget.MACHINE` since the new project is being created on the machine
- scaffold NewProject contrib and service, where the new project initialization will occur (selecting interpreter, creating new files, etc.)
- add configuration flag so the project wizard can be enabled via settings in release builds (setting ID `positron.projectWizardEnabled`)
    - project wizard will still be enabled by default in dev builds
    - update existing `IsDevelopmentContext` checks to also check this config setting

Next, I'll be leveraging the stored info in the NewProject contribution and service to initialize the project.

### QA Notes

The project wizard should be accessible in release builds by enabling the `positron.projectWizardEnabled` setting.

sharon-wang committed:
move env config update into reusable function

sharon-wang committed:
ensure the initial project config is saved

sharon-wang committed:
update comments and function docs

sharon-wang committed:
Update pythonEnvironmentStepUtils.ts to support undefined envType

sharon-wang committed:
store the new project configuration in the StorageService

sharon-wang committed:
add initial `PositronNewProjectContribution`

Using this new contribution, we can wait for workbench initialization and run the project initialization tasks.

sharon-wang committed:
use configuration flag to enable/disable Project Wizard

configuration id: `positron.useProjectWizard`

sharon-wang committed:
enable new project when config enabled or dev context

Add the flag to the Welcome page check as well.

nstrayer reviewed (approved):
Overall LGTM! 

I had a few philisophical questions about the feature flag + dev mode switching and comments. 

One thing I noticed in looking at the code was that some of the `useState()`s were doing extra computation. The code wasn't from this PR though so I couldn't figure out how to comment on it. 
In `pythonEnvironmentStep.tsx` for the `const [interpreterEntries, setInterpreterEntries]...` example it will rerun the `getPythonInterpreterEntries()` every re-render. (You can verify by putting a debug or `console.count()` at the top of that function. If you just switch to adding a lambda in front you avoid this. 
<img width="887" alt="image" src="https://github.com/posit-dev/positron/assets/6764693/e6e91883-ca13-477f-8956-06a00afb52a6">

sharon-wang committed:
comment on empty dependency array for `useEffect`

sharon-wang committed:
fix issue with clearing envType onEnvSetupSelected

Although envType is irrelevant for the Existing Environment env setup type, clearing it from the config means that if the user chooses New Environment after having chosen Existing Environment, the envType will be cleared and the user will have to re-select it.

The envType can be cleared if applicable when the project config is being constructed.

sharon-wang committed:
wrap `useState` initial values in arrow functions to avoid re-evaluation

Some of the `useState`s in pythonEnvironmentStep involve function calls that re-computed on every render. To avoid the unnecessary re-computation, wrap the initial values in arrow functions.

sharon-wang reviewed (commented):
@nstrayer Thank you for the `useState` extra computation tip! This is now in my React cheatsheet :)

I've modified a couple of the `useState`s in the pythonEnvironmentStep.tsx to use arrow functions to avoid the re-evaluation on each render.
